File: src/main/java/org/jukeboxmc/block/behavior/BlockCoralFan.java
package org.jukeboxmc.block.behavior;

import org.cloudburstmc.nbt.NbtMap;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.data.CoralColor;
import org.jukeboxmc.block.direction.RotationDirection;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.item.ItemType;
import org.jukeboxmc.item.behavior.ItemCoralFan;
import org.jukeboxmc.util.Identifier;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockCoralFan extends Block {

    public BlockCoralFan( Identifier identifier ) {
        super( identifier );
    }

    public BlockCoralFan( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public Item toItem() {
        return Item.<ItemCoralFan>create( ItemType.CORAL_FAN ).setCoralColor( this.getCoralColor() );
    }

    public void setCoralDirection( RotationDirection rotationDirection ) {
        this.setState( "coral_fan_direction", rotationDirection.ordinal() );
    }

    public RotationDirection getRotationDirection() {
        return this.stateExists( "coral_fan_direction" ) ? RotationDirection.values()[this.getIntState( "coral_fan_direction" )] : RotationDirection.EAST_WEST;
    }

    public BlockCoralFan setCoralColor( CoralColor coralColor ) {
        return this.setState( "coral_color", coralColor.name().toLowerCase() );
    }

    public CoralColor getCoralColor() {
        return this.stateExists( "coral_color" ) ? CoralColor.valueOf( this.getStringState( "coral_color" ) ) : CoralColor.BLUE;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockCoralFan.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.CoralFan
import org.jukeboxmc.api.block.data.CoralColor
import org.jukeboxmc.api.block.data.CoralFanDirection
import org.jukeboxmc.server.block.JukeboxBlock

class BlockCoralFan(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    CoralFan {

    override fun canPassThrough(): Boolean {
        return true
    }

   override fun getCoralColor(): CoralColor {
       return CoralColor.valueOf(this.getStringState("coral_color"))
   }

   override fun setCoralColor(value: CoralColor): BlockCoralFan {
       return this.setState("coral_color", value.name.lowercase())
   }

   override fun getCoralFanDirection(): CoralFanDirection {
       return CoralFanDirection.entries[this.getIntState("coral_fan_direction")]
   }

   override fun setCoralFanDirection(value: CoralFanDirection): BlockCoralFan {
       return this.setState("coral_fan_direction", value.ordinal)
   }
}